### ISSUE TITLE
Issues with reading AOD files

### ISSUE THREAD

**Issue by darousso** [scikit-hep/uproot4#210]
Hello,

Not sure if this is simply a feature that has not been implemented yet or me being incompetent and using this incorrectly, but I have been having some issues using uproot to read a test ATLAS AOD file I was wondering if you guys knew what may be causing this bug:

Thanks! And apologies if this is in the wrong place!
-David

(Version 0.1.2)

File: [myAOD.txt](https://github.com/scikit-hep/uproot4/files/5619826/myAOD.txt)
(I just changed the extension so github lets me upload it)

I ran:
```
import uproot4
file="myAOD.root"
xAOD=uproot4.open(file)
print(xAOD.keys())
Arrays=xAOD['CollectionTree;1'].arrays()
Array=xAOD['CollectionTree;1']['VrtSecInclusive_SecondaryVertices'].array()
```

and get for the last two lines:
```
---------------------------------------------------------------------------
NotNumerical                              Traceback (most recent call last)
~/.local/lib/python3.6/site-packages/uproot4/interpretation/identify.py in interpretation_of(branch, context, simplify)
    365             else:
--> 366                 from_dtype = _leaf_to_dtype(leaf).newbyteorder(">")
    367 

~/.local/lib/python3.6/site-packages/uproot4/interpretation/identify.py in _leaf_to_dtype(leaf)
     98     elif leaf.classname == "TLeafElement":
---> 99         return _ftype_to_dtype(leaf.member("fType"))
    100     else:

~/.local/lib/python3.6/site-packages/uproot4/interpretation/identify.py in _ftype_to_dtype(fType)
     67     else:
---> 68         raise NotNumerical()
     69 

NotNumerical: 

During handling of the above exception, another exception occurred:

RecursionError                            Traceback (most recent call last)
<ipython-input-8-f538788c4bba> in <module>
----> 1 xAOD['CollectionTree;1'].arrays()

~/.local/lib/python3.6/site-packages/uproot4/behaviors/TBranch.py in arrays(self, expressions, cut, filter_name, filter_typename, filter_branch, aliases, language, entry_start, entry_stop, decompression_executor, interpretation_executor, array_cache, library, how)
   1107             aliases,
   1108             language,
-> 1109             get_from_cache,
   1110         )
   1111 

~/.local/lib/python3.6/site-packages/uproot4/behaviors/TBranch.py in _regularize_expressions(hasbranches, expressions, cut, filter_name, filter_typename, filter_branch, keys, aliases, language, get_from_cache)
   3196         ):
   3197             if not isinstance(
-> 3198                 branch.interpretation,
   3199                 (
   3200                     uproot4.interpretation.identify.UnknownInterpretation,

~/.local/lib/python3.6/site-packages/uproot4/behaviors/TBranch.py in interpretation(self)
   2186             try:
   2187                 self._interpretation = uproot4.interpretation.identify.interpretation_of(
-> 2188                     self, {}
   2189                 )
   2190             except uproot4.interpretation.identify.UnknownInterpretation as err:

~/.local/lib/python3.6/site-packages/uproot4/interpretation/identify.py in interpretation_of(branch, context, simplify)
    438             out = uproot4.interpretation.objects.AsObjects(model_cls, branch)
    439             if simplify:
--> 440                 return out.simplify()
    441             else:
    442                 return out

~/.local/lib/python3.6/site-packages/uproot4/interpretation/objects.py in simplify(self)
    256                     header=False,
    257                     tobject_header=True,
--> 258                     original=self._model,
    259                 )
    260             except CannotBeStrided:

~/.local/lib/python3.6/site-packages/uproot4/model.py in strided_interpretation(cls, file, header, tobject_header, original)
   1021         versioned_cls = file.class_named(classname_decode(cls.__name__)[0], "max")
   1022         return versioned_cls.strided_interpretation(
-> 1023             file, header=header, tobject_header=tobject_header
   1024         )
   1025 

<dynamic> in strided_interpretation(cls, file, header, tobject_header, original)

... last 2 frames repeated, from the frame below ...

~/.local/lib/python3.6/site-packages/uproot4/model.py in strided_interpretation(cls, file, header, tobject_header, original)
   1021         versioned_cls = file.class_named(classname_decode(cls.__name__)[0], "max")
   1022         return versioned_cls.strided_interpretation(
-> 1023             file, header=header, tobject_header=tobject_header
   1024         )
   1025 

RecursionError: maximum recursion depth exceeded in comparison

```

**Comment by jpivarski** [comment]
I can reproduce it with:

```python
>>> file = uproot.open("uproot-issue-210.root")
>>> file["CollectionTree/VrtSecInclusive_SecondaryVertices"].interpretation
```

It's not in the array-reading, it's in determining the interpretation. I'm looking into it.

**Comment by jpivarski** [comment]
I've fixed the infinite recursion, but the branch still can't be read. At least you'll be able to `show` that TTree and read the other branches.

The issue is that the type technically contains itself:

```python
>>> f = uproot.open("uproot-issue-210.root")
>>> f.file.streamer_named("DataVector<xAOD::Vertex_v1>").show()
DataVector<xAOD::Vertex_v1> (v1)
    This: DataVector<xAOD::Vertex_v1> (TStreamerSTL)
```

However, my interpretation of this as a class with one member named "`This`" whose type is the class is probably wrong. The single member named "`This`" is probably intended to be a signal to ROOT to interpret this some other way. Maybe "`DataVector`" is a typedef for "`std::vector`"? If so, then this is the wrong interpretation.

I checked to see if Uproot 3 had a special interpretation for it that simply hasn't been ported over to Uproot 4: it doesn't.

This will probably have to be revisited. However, the first thing that happened when I tried reading this is that it complained about "memberwise splitting" not being implemented, so I'll probably need to do #38 first, at least for `std::vectors` (if that's what `DataVector` is). #209 also needs memberwise `std::vector` interpretation.
